### ISSUE TITLE
[SYCL][CUDA] Incorrect work with constant memory

### ISSUE THREAD

**Issue by anton-v-gorshkov** [intel/llvm#1819]
[main.cpp.txt](https://github.com/intel/llvm/files/4734436/main.cpp.txt)

After migration of Rodinia-Heartwall benchmark from CUDA to DPC++ it was found that Codeplay DPC++ compiler may work with constant memory incorrectly - constant data passed to the kernel may be invalid.

Small reproducer is in attachment.

How to reproduce:
```
$ source /opt/intel/inteloneapi/dpcpp-ct/latest/env/vars.sh
$ clang++ -O0 -g -std=c++11 -fsycl -fsycl-unnamed-lambda -fsycl-targets=nvptx64-nvidia-cuda-sycldevice main.cpp -o test -Xsycl-target-backend=nvptx64-nvidia-cuda-sycldevice --cuda-gpu-arch=sm_75 -lsycl -lOpenCL
$ export SYCL_BE=PI_CUDA
$ ./test
Device: GeForce RTX 2070
Expected Output: 123, 321
Real Output: 123, 1 # INVALID
```

I've used the latest compiler from June 04.

For Intel Graphics it works fine:
```
$ clang++ -O0 -g -std=c++11 -fsycl -fsycl-unnamed-lambda main.cpp -o test -lsycl -lOpenCL
$ unset SYCL_BE
$ ./test
Device: Intel(R) Gen9 HD Graphics NEO
Expected Output: 123, 321
Real Output: 123, 321 # VALID
```

**Comment by Ruyk** [comment]
Can you confirm the problem only occurs with constant memory? I have modified locally the test case to use non-constant memory but get the same error.

**Comment by Ruyk** [comment]
I think this could be related to https://github.com/intel/llvm/pull/1841

**Comment by Ruyk** [comment]
It also seems the problem disappears when using -O2 instead of -O0

**Comment by anton-v-gorshkov** [comment]
> It also seems the problem disappears when using -O2 instead of -O0

True

**Comment by anton-v-gorshkov** [comment]
> Can you confirm the problem only occurs with constant memory? I have modified locally the test case to use non-constant memory but get the same error.

The issue was discovered for constant memory only - but if you could attach your version of reproducer, I can check it on my machine as well.

**Comment by bader** [comment]
Can't reproduce this problem with the latest build.

>  SYCL_BE=PI_CUDA ./test
> 
> WARNING: The legacy environment variables SYCL_BE and SYCL_DEVICE_TYPE are deprecated. Please use SYCL_DEVICE_FILTER instead. For details, please refer to https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md
> 
> Device: TITAN RTX
> Expected Output: 123, 321
> Real Output: 123, 321

It was (probably) addressed by #1841.
